crates/miden-objects/src/batch/proposed_batch.rs

```diff
@@ -0,0 +1,508 @@
+use alloc::{
+    collections::{btree_map::Entry, BTreeMap, BTreeSet},
+    sync::Arc,
+    vec::Vec,
+};
+
+use crate::{
+    account::AccountId,
+    batch::{BatchAccountUpdate, BatchId, BatchNoteTree},
+    block::{BlockHeader, BlockNumber},
+    errors::BatchProposeError,
+    note::{NoteHeader, NoteId, NoteInclusionProof},
+    transaction::{
+        ChainMmr, InputNoteCommitment, InputNotes, OutputNote, ProvenTransaction, TransactionId,
+    },
+    MAX_ACCOUNTS_PER_BATCH, MAX_INPUT_NOTES_PER_BATCH, MAX_OUTPUT_NOTES_PER_BATCH,
+};
+
+/// A proposed batch of transactions with all necessary data to validate it.
+///
+/// See [`ProposedBatch::new`] for what a proposed batch expects and guarantees.
+///
+/// This type is fairly large, so consider boxing it.
+#[derive(Debug, Clone)]
+pub struct ProposedBatch {
+    /// The transactions of this batch.
+    transactions: Vec<Arc<ProvenTransaction>>,
+    /// The header is boxed as it has a large stack size.
+    block_header: BlockHeader,
+    /// The chain MMR used to authenticate:
+    /// - all unauthenticated notes that can be authenticated,
+    /// - all block hashes referenced by the transactions in the batch.
+    chain_mmr: ChainMmr,
+    /// The note inclusion proofs for unauthenticated notes that were consumed in the batch which
+    /// can be authenticated.
+    unauthenticated_note_proofs: BTreeMap<NoteId, NoteInclusionProof>,
+    /// The ID of the batch.
+    id: BatchId,
```

Comment:
I would maybe mention that batch ID is a cryptographic commitment to the transactions in the batch.